File: JAVA/Pertemuan 01/src/id/ac/uph/pbo/pertemuan01/ProsesAbstraksi.java
package id.ac.uph.pbo.pertemuan01;

public class ProsesAbstraksi {

    public static void main(String[] args) {

        System.out.println("╔══════════════════════════════════════════════╗");
        System.out.println("║   PROSES ABSTRAKSI – Satu Objek, Dua Konteks ║");
        System.out.println("╚══════════════════════════════════════════════╝\n");

        System.out.println("Orang yang sama di dunia nyata: 'Sari Wulandari'");
        System.out.println("Diabstraksikan BERBEDA sesuai konteks sistem:\n");

        // ── Konteks 1: Sistem Akademik ───────────────────────────────────────
        System.out.println("── KONTEKS 1: Sistem Akademik ──");
        MahasiswaAkademik sariAkademik =
                new MahasiswaAkademik("Sari Wulandari", "220301002",
                                       3.75, 3, "Sistem Informasi");
        sariAkademik.tampilkanInfo();
        sariAkademik.ambilMataKuliah("Pemrograman Berorientasi Objek");
        sariAkademik.hitungIPK(new double[]{3.5, 4.0, 3.75, 3.8});

        // ── Konteks 2: Sistem Perpustakaan ───────────────────────────────────
        System.out.println("\n── KONTEKS 2: Sistem Perpustakaan ──");
        MahasiswaPerpustakaan sariPustaka =
                new MahasiswaPerpustakaan("Sari Wulandari", "220301002");
        sariPustaka.pinjamBuku("Object-Oriented Software Construction");
        sariPustaka.pinjamBuku("Head First Java");
        sariPustaka.kembalikanBuku(false);   // tepat waktu
        sariPustaka.kembalikanBuku(true);    // terlambat
        sariPustaka.tampilkanInfo();

        System.out.println("\n──────────────────────────────────────────────────");
        System.out.println("KESIMPULAN: Abstraksi = pilih atribut yang RELEVAN");
        System.out.println("           sesuai kebutuhan sistem. Detail yang tidak");
        System.out.println("           relevan (warna baju, hobi) → DIABAIKAN.");
        System.out.println("──────────────────────────────────────────────────");
    }
}

File: PYTHON/Pertemuan 01/ProsesAbstraksi.py
import sys
sys.stdout.reconfigure(encoding='utf-8')
from MahasiswaAkademik import MahasiswaAkademik
from MahasiswaPerpustakaan import MahasiswaPerpustakaan

def main():
    print("╔══════════════════════════════════════════════╗")
    print("║   PROSES ABSTRAKSI – Satu Objek, Dua Konteks ║")
    print("╚══════════════════════════════════════════════╝\n")

    print("Orang yang sama di dunia nyata: 'Sari Wulandari'")
    print("Diabstraksikan BERBEDA sesuai konteks sistem:\n")

    # ── Konteks 1: Sistem Akademik ───────────────────────────────────────
    print("── KONTEKS 1: Sistem Akademik ──")
    sari_akademik = MahasiswaAkademik("Sari Wulandari", "220301002", 3.75, 3, "Sistem Informasi")
    sari_akademik.tampilkan_info()
    sari_akademik.ambil_mata_kuliah("Pemrograman Berorientasi Objek")
    sari_akademik.hitung_ipk([3.5, 4.0, 3.75, 3.8])

    # ── Konteks 2: Sistem Perpustakaan ───────────────────────────────────
    print("\n── KONTEKS 2: Sistem Perpustakaan ──")
    sari_pustaka = MahasiswaPerpustakaan("Sari Wulandari", "220301002")
    sari_pustaka.pinjam_buku("Object-Oriented Software Construction")
    sari_pustaka.pinjam_buku("Head First Java")
    sari_pustaka.kembalikan_buku(False)   # tepat waktu
    sari_pustaka.kembalikan_buku(True)    # terlambat
    sari_pustaka.tampilkan_info()

    print("\n──────────────────────────────────────────────────")
    print("KESIMPULAN: Abstraksi = pilih atribut yang RELEVAN")
    print("           sesuai kebutuhan sistem. Detail yang tidak")
    print("           relevan (warna baju, hobi) → DIABAIKAN.")
    print("──────────────────────────────────────────────────")

if __name__ == "__main__":
    main()

